### ISSUE TITLE
When ZOWE_HOST_NAME cannot be resolved default to ZOWE_IP_ADDRESS in instance.env

### ISSUE THREAD

**Issue by Joe-Winchester** [zowe/zowe-install-packaging#1279]
**Describe the bug**
On some systems the ZOWE_HOST_NAME cannot be resolved externally.  It is used in the configure.sh scripts for the desktop exolorers, and this causes 404 errors

**Expected behavior**
When the ZOWE_HOST_NAME cannot be resolved if none is specified as an override then default to the ZOWE_IP_ADDRESS so that the instance.env file ZOWE_HOST_NAME is the ZOWE_IP_ADDRESS

This logic will be zowe-init.sh.  

**Screenshots**
When parsing the results of the dig command on the system in error is shown below

![image (12)](https://user-images.githubusercontent.com/28100302/79768326-8911e700-8322-11ea-9fdf-2e27c6cfdaf1.png)

The result of running zowe-init.sh on the system in error is shown belos
![image (13)](https://user-images.githubusercontent.com/28100302/79768306-80b9ac00-8322-11ea-80f1-2f13d1640260.png)

**Additional context**
This problem will resolve https://github.com/zowe/explorer-jes/pull/169

**Comment by Joe-Winchester** [comment]
Behavior should be
1) If the ZOWE_HOST_NAME cannot be located prompt, and if one if not supplied then make the ZOWE_NAME_NAME be the IP address.  The IP address logic is correct as this identifies it correctly on the zD&T system where the problem is occuring.

Currently instance.env has the two variables this way around, e.g.
```
ZOWE_EXPLORER_HOST=S0W1.DAL-EBIS-IHOST.COM
ZOWE_IP_ADDRESS=192.86.32.67
```
so in the case where the ZOWE_EXPLORER_HOST cannot be resolved it will become
```
ZOWE_EXPLORER_HOST=192.86.32.67
ZOWE_IP_ADDRESS=192.86.32.67
```
If you are feeling confident then an optimization would be to write the `instance.env` variables
```
ZOWE_IP_ADDRESS=192.86.32.67
ZOWE_EXPLORER_HOST=$ZOWE_IP_ADDRESS
```
for un diggable host names (i.e. IP above HOST_NAME)

**Comment by John-A-Davies** [comment]
The screenshot shows the user not entering the ZOWE_IP_ADDRESS, instead of not entering the ZOWE_EXPLORER_HOST, so it would be good to capture both together to be sure we test cases where either or both are missing.  

I don't think `zowe-configure-instance.sh` scans  `instance.env` before it tries to determine host and ip, and it only updates them in  `instance.env` if they are not already present.
ZOWE_IP_ADDRESS and ZOWE_EXPLORER_HOST are entangled, so they need to be treated together, which might mean some rework of the processing order in `zowe-init.sh`.

**Comment by John-A-Davies** [comment]
That's fine for the system above, where it says "dig could not find IP of hostname", i.e. dig fails.  But I've also added a check for when dig succeeds, and it says "ping S0W1.DAL-EBIS.IHOST.COM did not match stated IP address", which is normal for a ZD&T system.

**Comment by John-A-Davies** [comment]
We should perhaps document our solution to this issue.  Suggested wording to add to  
` docs-site/docs/user-guide/configure-instance-directory.md `

When you run `zowe-configure-instance.sh`, it uses `ping` and `dig` internally to derive the values for ZOWE_IP_ADDRESS and ZOWE_EXPLORER_HOST.  If it cannot determine these,  it prompts the user for their values.  If ZOWE_EXPLORER_HOST does not resolve to ZOWE_IP_ADDRESS, it sets ZOWE_EXPLORER_HOST to the same value as ZOWE_IP_ADDRESS.

**Comment by jackjia-ibm** [comment]
@Joe-Winchester @John-A-Davies  I think this request is problematic and we need to rethink or revert some of the changes we made. Basic line is we should respect what the user defined in `ZOWE_EXPLORER_HOST` and try not to modify it.

Based on the domain configuration and nameserver configuration of the z/OS server, on server we may not be able to find out the resolution of the domain, but that doesn't mean the domain is not valid or not accessible from external. We can still keep some guess works, but only when the user didn't define `ZOWE_EXPLORER_HOST`.

**Comment by jackjia-ibm** [comment]
A better approach I think we should separate HOST to `ZOWE_EXTERNAL_HOST` and `ZOWE_INTERNAL_HOST`.

The `ZOWE_EXTERNAL_HOST` will be used to access gateway, desktop and explorer UI iframe plugins. This is external domain, we don't need to validate on z/OS. That's the domain/url eventually the user will be put into browser address bar. It should be accessible/resolved from the client side.

The `ZOWE_INTERNAL_HOST` will be used to internal communication between services and gateway. This can usually be defined as `localhost`, or the `$(hostname)`. This can be same as `ZOWE_EXTERNAL_HOST` if the z/OS server has nameserver defined or has the domain defined in `/etc/hosts`. If we use `localhost` here, the z/OS server should be able to resolve `localhost` to 127.0.0.1.

With this change, we have possibility to not expose ports (like explorer APIs, UI plugins, API catalog) to external but just bind them to 127.0.0.1.

**Comment by John-A-Davies** [comment]
On ZD&T systems, the ZOWE_EXTERNAL_HOST would be commuted to the IP address, while the ZOWE_INTERNAL_HOST would remain at S0W1.DAL-EBIS.IHOST.COM, so the latter would be used in `zowe-setup-certificates.sh`.

**Comment by jackjia-ibm** [comment]
Yes, if the user choose to configure ZOWE_EXTERNAL_HOST to be IP address for zD&T, that's their choice. We just need to make sure we don't put `dns:<ip-address>` into certificate SAN which will cause a failure.

We may need both external and internal host name in the self-signed certificate we generated because we use it to encrypt both `browser <-> z/os` communication and `gateway <-> other services` communication. Or we can choose to generate 2 sets of certificates for 2 purposes.

**Comment by John-A-Davies** [comment]
@Joe-Winchester Do we need to revise the solution to this issue, to include the external vs. external HOSTNAME and IP concept?

**Comment by Joe-Winchester** [comment]
My understanding - currently `instance.env` holds the two variables  
```
ZOWE_HOST_HAME=
ZOWE_IP_ADDRESS=
```
The change that @John-A-Davies made was that if during the calculation of the defaults, if it is detected that the value that is going to be set into ZOWE_HOST_NAME isn't actually externally accessible (it comes from running `hostname`) then rather than use it, just set it to be a numeric IP address.  
However as @jackjia-ibm points out depending on who uses this can run into problems.  For the self-signed certificate generation I think the values used come from `zowe-setup-certificates.env`that has its own `HOSTNAME=` and `IPADDRESS=`  but I don't know how these get default values (in my install they are blank).    @OnnoVdT and @stevenhorsman - what are you suggestions ?

**Comment by Joe-Winchester** [comment]
@jackjia-ibm https://github.com/zowe/zowe-install-packaging/issues/1279#issuecomment-626767571 I think that we do always respect what the user types into `ZOWE_HOST_NAME` in `instance.env` - it's just the default value that is set to be an IP address if we know it can't be resolved.  @John-A-Davies - can you confirm that the only change was to generate a different default and not to actually splat or ignore what the user has entered in instance.env ?

**Comment by stevenhorsman** [comment]
It looks like from #1350 that the behaviour isn't working properly, but the way it is supposed to work is to use the values in `zowe-setup-certificates.env` if supplied and otherwise get them from `zowe-init.sh`. This means that the default's behaviour has now been changed to set the hostname to be an ip address now.
If zowe-setup-certificates worked correctly this would result in a failure in the cases that the user hasn't specified `HOSTNAME=` in `zowe-setup-certificates.env` and the new check John added fails resulting in the certificates being generated with a numeric hostname which doesn't work.

**Comment by John-A-Davies** [comment]
@jackjia-ibm The change in `zowe-init.sh` asks the user for ZOWE_EXPLORER_HOST and ZOWE_IP_ADDRESS if they are blank or there is a problem resolving them.  Finally, it sets ZOWE_EXPLORER_HOST=ZOWE_IP_ADDRESS if ZOWE_EXPLORER_HOST does not resolve to ZOWE_IP_ADDRESS.

**Comment by John-A-Davies** [comment]
@jackjia-ibm After `zowe-init.sh` ends, the values of ZOWE_EXPLORER_HOST and ZOWE_IP_ADDRESS it has calculated will be appended to `instance.env` ... *unless they were already present*.  This 'unless' part has not changed.

**Comment by jackjia-ibm** [comment]
I believe this is the root cause of https://github.com/zowe/zowe-install-packaging/issues/1350.

In `bin/zowe-setup-certificates.sh`, line 62 ~ 64:

```
ZOWE_EXPLORER_HOST=${HOSTNAME}
ZOWE_IP_ADDRESS=${IPADDRESS}
. ${ZOWE_ROOT_DIR}/bin/zowe-init.sh -s
```

If the user configured `HOSTNAME` as their correct and external domain name, for example, `tvt6019.svl.ibm.com`, after `zowe-init.sh`, `ZOWE_EXPLORER_HOST` is replaced to IP address, and then be put into `SAN=dns:${ZOWE_EXPLORER_HOST}` on line 91. This caused `apiml_cm.sh` failed on generating certificates.

That's why I think overwrite `ZOWE_EXPLORER_HOST` with `ZOWE_IP_ADDRESS` is not right because the domain check and also IP check on z/OS is not trustworthy. What you can detect on z/OS may not be what the user eventually using to connect to the service.

**Comment by stevenhorsman** [comment]
Yep - thanks for investigating this more Jack - it looks like the PR caused the regression #1350 as well, so we need to get that fixed ASAP as well!

**Comment by John-A-Davies** [comment]
@NakulManchanda Please let me know if your fix makes this one redundant.  The fix is in PR #1372

**Comment by John-A-Davies** [comment]
Fixed by PR #1371
